static/js/formatters-internal.js

```diff
@@ -311,164 +311,117 @@ export function image(simpleOrComplexImage = {}, desiredSize = '200x', atLeastAs
+  // only height is provided
+  if (!desiredWidthProvided && desiredHeightProvided) {
+    formatOptions.push(`height=${desiredHeight}`);
 
-    fullSizeWidth = Number.parseInt(fullSizeDims[0]);
-    fullSizeHeight = Number.parseInt(fullSizeDims[1]);
+    return `/${formatOptions.join(',')}`;
   }
 
-  let desiredDims = desiredSize.split('x');
-  let formatOptions = [];
-
-  if (desiredDims[0] !== '') {
-    const desiredWidth = Number.parseInt(desiredDims[0]);
-    if (Number.isNaN(desiredWidth)) {
-      throw new Error("Invalid width specified");
-    }
-
+  // both dimensions are provided
+  if (atLeastAsLarge && !!originalRatio) {
+    formatOptions.push(`width=${Math.max(desiredWidth, Math.round(desiredHeight * originalRatio))}`);
+    formatOptions.push(`height=${Math.max(desiredHeight, Math.round(desiredWidth / originalRatio))}`);
+  } else {
     formatOptions.push(`width=${desiredWidth}`);
-  } else if (!atLeastAsLarge && fullSizeWidth) {
-    formatOptions.push(`width=${fullSizeWidth}`);
-  }
-
-  if (desiredDims[1] !== '') {
-    const desiredHeight = Number.parseInt(desiredDims[1]);
-    if (Number.isNaN(desiredHeight)) {
-      throw new Error("Invalid height specified");
-    }
-
     formatOptions.push(`height=${desiredHeight}`);
-  } else if (!atLeastAsLarge && fullSizeHeight) {
-    formatOptions.push(`height=${fullSizeHeight}`);
   }
 
-  formatOptions.push(`fit=${atLeastAsLarge ? 'cover' : 'contain'}`);
-
-  const urlWithOptions = image.url + `/${formatOptions.join(',')}`;
-
-  return _replaceUrlHost(urlWithOptions, 'dyn.eu.mktgcdn.com');
-}
-
-/**
- * Splits a string into two parts at the specified index.
- *
- * @param {string} str The string to be split
- * @param {number} index The index at which to split the string
- * @returns {Array<string>} The two parts of the string after splitting
- */
-function _splitStringOnIndex(str, index) {
-  return [str.slice(0, index), str.slice(index)];
+  return `/${formatOptions.join(',')}`;
 }
 
 /**
- * Replaces the current host of a url with the specified host.
+ * Given a photo image url path, remove the trailing extension.
  *
- * @param {string} url The url whose host is to be changed
- * @param {string} host The new host to change to
- * @returns {string} The url updated with the specified host
+ * @param {string} imageUrlPath Image's url path (e.g.
+ *   '/p/mFsjqWGQEOMQGNoNIcnq61JtdSGiCs/225x225.jpg',
+ *   '/p-sandbox/mFsjqWGQEOMQGNoNIcnq61JtdSGiCs/225x225.jpg')
+ * @returns {string} A canonicalized image url path (e.g.
+ *   '/p/mFsjqWGQEOMQGNoNIcnq61JtdSGiCs',
+ *   '/p-sandbox/mFsjqWGQEOMQGNoNIcnq61JtdSGiCs')
  */
-function _replaceUrlHost(url, host) {
-  const splitUrl = url.split('://');
-  const urlAfterHost = splitUrl[1].slice(splitUrl[1].indexOf('/'));
-  return splitUrl[0] + '://'  + host + urlAfterHost;
+function _removePhotoImageUrlExtension(imageUrlPath) {
+  return imageUrlPath.replace(/(\/[0-9]+x[0-9]+\.[a-z]+)|(\/)$/, '');
```

Comment:
is there a reason we check for the image dimensions in this regex? could we just split on the last index of `.` instead?

Reply:
Do you mean the last index of `/`?

I almost did comment that photo urls ending with `/original` do exist (no width/height and no extension), but you should never see one in entity data so it didn't seem that relevant. This should be safe in practice, but just getting the directory name would also work.

Reply:
Seems like using a regex is safer just in case we get a url without width/height and extension? Keeping it for now

Reply:
ah yes, I meant the last `/`.

could a photo url ever end in `/`? and maybe I'm misunderstanding the last part of the regex, but why would we want to remove any arbitrary last `/` if it doesn't match the `/200x200.jpg` format? also, do we get the height/width information through some other means now if we remove it from the url before doing the dimension parsing? is it guaranteed to always be a part of the image object?

Reply:
I just wanted to be sure there's no trailing `/` before we append the format options

Reply:
> is it guaranteed to always be a part of the image object?

Size should always be returned for image objects. There have been bugs over the years where the metadata didn't get recorded, but to the best of my knowledge they've been fixed. So, width/height is _expected_ to be on any image-type entity field-value at least.

(It's also possible to upload an image to a generic File-type field; the entity value wouldn't have image metadata in this case, but also the URL wouldn't necessarily adhere to the same format anyway)

Reply:
would this regex remove `/200x200.jpg/` if there was a trailing slash after the dimensions and extension?

and @jfancher-yext, is it common to upload an image to a generic file-type field? and do you know what format the URL would have? trying to determine if this is something we need support for, and if so, how much work it would be

Reply:
It is not common in my experience, though the [docs](https://hitchhikers.yext.com/docs/content/image-storing-and-serving/) do mention it as a possibility (n.b. these docs are slightly incorrect otherwise, will be looking to get them corrected as part of this project).

Images as Files would be the same as all images are in the EU except with the US domain -- `a.mktgcdn.com/f[-env]/id/hash.ext`. If this is supported, ideally the id segment would be optional, i.e., also allow `a.mktgcdn.com/f[-env]/hash.ext`. We would like to remove that segment; due to hard-to-change dependencies it's not going to happen any time soon, so it's not urgent, but in any case to make the dynamic urls for `/f/` inputs it's a simple algorithm: replace `a.` with `dyn.` and add `/width=x,height=y,etc` to the end; no need to worry about the exact structure beyond that.

(As we consolidate the two systems we're going to be changing this so that images are stored as a `/p/...` URL in the US even if they comes in via the generic File API, but existing file images won't move)

Reply:
Got it. So it sounds like the URL formatting should already be handled by the changes in this PR. But, if the entity has no image metadata and the URL doesn't include the dimensions, is there any way to get the original image size?

Reply:
If it doesn't have metadata (which again would be a bug if it happened for an Image-type field, but expected for a File-type field) then no, it's not possible short of actually requesting the image.

Reply:
Ah ok, so sounds like it wouldn't really be possible (or at least pretty difficult) for us to support this fully. Since it's not that common, I think we can continue without supporting it unless we get a feature request for it